// @upstash/qstash/esm/pkg/client/messages.js
export class Messages {
    constructor(http) {
        Object.defineProperty(this, "http", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.http = http;
    }
    /**
     * Get a message
     */
    async get(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "messages", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * List your messages
     */
    async list(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "messages"],
            headers: { "Content-Type": "application/json" },
            query: { cursor: req?.cursor },
        });
    }
    /**
     * List logs from a message
     */
    async logs(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "messages", req.id, "logs"],
            headers: { "Content-Type": "application/json" },
            query: { cursor: req.cursor },
        });
    }
    /**
     * List tasks for a message
     */
    async tasks(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "messages", req.id, "tasks"],
            headers: { "Content-Type": "application/json" },
            query: { cursor: req.cursor },
        });
    }
    /**
     * Cancel a topic by name or id.
     */
    async delete(req) {
        return await this.http.request({
            method: "DELETE",
            path: ["v1", "messages", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
}
