// @upstash/qstash/script/pkg/client/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoints = void 0;
class Endpoints {
    constructor(http) {
        Object.defineProperty(this, "http", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.http = http;
    }
    /**
     * Create a new endpoint with the given name.
     */
    async create(req) {
        return await this.http.request({
            method: "POST",
            path: ["v1", "endpoints"],
            body: JSON.stringify(req),
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * Get a list of all endpoints.
     */
    async list() {
        return await this.http.request({
            method: "GET",
            path: ["v1", "endpoints"],
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * Get a single endpoint.
     */
    async get(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "endpoints", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * Update a endpoint
     */
    async update(req) {
        return await this.http.request({
            method: "PUT",
            path: ["v1", "endpoints", req.id],
            body: JSON.stringify({ url: req.url }),
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * Delete a endpoint.
     */
    async delete(req) {
        return await this.http.request({
            method: "DELETE",
            path: ["v1", "endpoints", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
}
exports.Endpoints = Endpoints;
