// @upstash/qstash/script/pkg/client/schedules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedules = void 0;
class Schedules {
    constructor(http) {
        Object.defineProperty(this, "http", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.http = http;
    }
    /**
     * Get a schedule
     */
    async get(req) {
        return await this.http.request({
            method: "GET",
            path: ["v1", "schedules", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * List your schedules
     */
    async list() {
        return await this.http.request({
            method: "GET",
            path: ["v1", "schedules"],
            headers: { "Content-Type": "application/json" },
        });
    }
    /**
     * Delete a schedule
     */
    async delete(req) {
        return await this.http.request({
            method: "DELETE",
            path: ["v1", "schedules", req.id],
            headers: { "Content-Type": "application/json" },
        });
    }
}
exports.Schedules = Schedules;
